bip-00nn.mediawiki

```diff
@@ -0,0 +1,120 @@
+
+When executed <code>OP_CHECKSIGEX</code> pops an integer from the top of the stack. This integer encodes the binary options which determine how <code>OP_CHECKSIGEX</code> verifies signatures. The following table shows the possible flags at the time of writing:
+
+{| class="wikitable"
+!colspan="6"|OP_CHECKSIGEX flags
+|-
+! Name
+! Value
+! Description
+|-
+| <code>SCRIPT_CHECKSIGEX_VERIFY</code>
+| 2^0
+| Verify after checking validity and fail script if verification failed
+|-
+| <code>SCRIPT_CHECKSIGEX_MULTI</code>
+| 2^1
+| Mimic <code>OP_CHECKMULTISIG</code> if set otherwise mimic <code>OP_CHECKSIG</code>
+|-
+| <code>SCRIPT_CHECKSIGEX_NORMALIZE</code>
+| 2^2
+| Normalize input hashes before checking signature
+|}
+
+A combination of <code>SCRIPT_CHECKSIGEX_VERIFY</code> and </code>SCRIPT_CHECKSIGEX_MULTI</code> recreates the semantics of <code>OP_CHECKSIG</code>, <code>OP_CHECKSIGVERIFY</code>, <code>OP_CHECKMULTISIG</code> and <code>OP_CHECKMULTISIGVERIFY</code>. Refer to the developer documentation<ref>[[https://bitcoin.org/en/developer-guide#p2pkh-script-validation|Developer Documentation - P2PKH Script Validation]]</ref> for their semantics.
+
+The existing <code>OP_CHECKMULTISIG</code> and <code>OP_CHECKMULTISIGVERIFY</code> have a bug<ref>[[https://bitcoin.org/en/developer-guide#multisig|Developer Documentation - Multisig]]</ref> that pops one argument too many from the stack. This bug is not reproduced in the implementation of OP_CHECKSIGEX, so the canonical solution of pushing a dummy value onto the stack is not necessary.
+
+<code>SCRIPT_CHECKSIGEX_NORMALIZE</code> indicates that before computing the signaturehash, i.e., the hash that is signed, the transaction should be normalized by replacing all transaction IDs in the inputs by their normalized variants. The normalized transction IDs are computed as described in the next section. This normalization step is performed both when creating the signatures as well as when checking the signatures.
+
+In order to make use of the normalization feature, this BIP introduces two new standard transaction types
+
+    OP_DUP OP_HASH160 <pubkey hash> OP_EQUALVERIFY OP_4 OP_CHECKSIGEX
+
+and 
+
+    <num sigs> <pubkeys> <numpubkeys> OP_6 OP_CHECKSIGEX
+
+The first is the equivalent of the pay-to-pubkeyhash script with normalization (<code>OP_4 = SCRIPT_CHECKSIGEX_NORMALIZE</code>), while the second is the equivalent of the pay-to-multisig also with normalization (<code>OP_6 = SCRIPT_CHECKSIGEX_NORMALIZE | SCRIPT_CHECKSIGEX_MULTI</code>). The format of the <code>scriptSig</code> corresponds to the original version, with exception of omitting the dummy <code>OP_0</code> in the multisig case.
+
+=== Tracking Normalized Transaction IDs ===
```

Comment:
There is no change for bloomfilters in that we maintain the existing semantics: if a bloomfilter matches a transaction that is modified in flight it might not match the modified transaction and it may miss future transactions building on it. We could add the normalized outpoint to the bloomfilter as well to ensure that we start forwarding modified transactions and transactions dependent on modified transactions as well, but that might be better suited to a separate spec.